### 图神经网络环境配置，参见 https://www.bilibili.com/video/BV1184y1x71H?p=11

In [1]:
import os
import numpy as np
import pandas as pd
import os
import pandas as pd
from scipy.io import loadmat
import networkx as nx
import matplotlib.pyplot as plt
import torch
from torch_geometric.data import InMemoryDataset, Data
from tqdm import tqdm
from torch_geometric.loader import DataLoader
from sklearn.metrics import precision_score, f1_score, recall_score, accuracy_score
from torch_geometric.nn import TopKPooling,SAGEConv, GCNConv, SAGPooling
from torch_geometric.nn import global_mean_pool as gap, global_max_pool as  gmp
from torch.nn import Linear,BatchNorm1d, ReLU
import torch.nn.functional as F
from imblearn.under_sampling import RandomUnderSampler

device = torch.device("mps" if torch.cuda.is_available() else "cpu")
device

device(type='cpu')

## 参数

In [2]:
epoch_num = 30  # 训练轮数
lr = 0.001      # 学习率
bs = 64         # 批次大小
isCause = True # 是否加入因果关系

dataset_root = '../Data' # 数据存放目录

## 因果关系的加入
注意：节点必须从0开始标号

In [3]:
def get_edge_index_and_edge_attr(ex_type):
    # source_nodes->target_nodes对应着一条有向边，而edge_attr为此条有向边的权值,无向图代表两个方向都有边
    source_nodes = [0, 1, 0, 4, 0, 7, 1, 2, 2, 3, 4, 5, 5, 6, 7, 8, 8, 9, 8, 14, 8, 19, 9, 10, 10, 11, 11, 12, 12, 13, 14, 15, 15, 16, 16, 17, 17, 18, 19, 20, 20, 21]
    target_nodes = [1, 0, 4, 0, 7, 0, 2, 1, 3, 2, 5, 4, 6, 5, 8, 7, 9, 8, 14, 8, 19, 8, 10, 9, 11, 10, 12, 11, 13, 12, 15, 14, 16, 15, 17, 16, 18, 17, 20, 19, 21, 20]
    edge_attr = None
    if isCause:
        # # one leg stand
        # if ex_type == 1:
        #     source_nodes = [0, 1, 2, 4, 6, 7, 8, 8 , 8 , 9, 10, 11, 12, 15, 15, 16, 17, 19, 19, 21]
        #     target_nodes = [4, 0, 3, 5, 5, 0, 7, 9, 14, 10, 11, 12, 13, 14, 16, 17, 18, 8 , 20, 20]
        #     edge_attr = None
        if ex_type == 3:
            source_nodes = [4, 1, 3, 5, 5, 7, 8, 8 , 8 , 9, 11, 11, 14, 16, 16, 19, 20]
            target_nodes = [0, 0, 2, 4, 6, 0, 7, 9, 14, 10, 10, 12, 15, 15, 17, 8 , 19]
            edge_attr = None
    edge_index = torch.tensor([source_nodes, target_nodes], dtype=torch.long)
    return edge_index, edge_attr

## 数据处理部分
注意：如果图结构发生改变，请删除数据根目录/processed/下的缓存文件，以重新进行数据处理

In [4]:
# 这个类继承自InMemoryDataset，专门处理中小型数据集
class BinaryDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None, pre_filter=None):
        super().__init__(root, transform, pre_transform, pre_filter)
        self.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return []
    # 如果根目录下的processed文件夹下已经尤处理好的图数据，则直接读取
    @property
    def processed_file_names(self):
        if isCause:
            return ['cause.dataset']
        else:
            return ['base.dataset',]
    def download(self):
        # Download to `self.raw_dir`.
        pass

    def process(self):
        """
        Load Data Seperate All data in Protective and Non-Prtective Data
        """
        # Assuming the 'Data' folder is in the current working directory
        data_folders = [dataset_root]

        # Initialize lists to store the separated dataframes
        protective_dfs = []
        non_protective_dfs = []

        # Loop through each data folder
        for data_folder in data_folders:
            # List all .mat files in the current data folder
            mat_files = [f for f in os.listdir(data_folder) if f.endswith('.mat')]
            # Load each mat file
            for mat_file in mat_files:
                # Construct the full path to the .mat file
                mat_path = os.path.join(data_folder, mat_file)
                # Load the .mat file
                mat_data = loadmat(mat_path)
                # Convert the data into a pandas dataframe
                df = pd.DataFrame(mat_data['data'])
                # Select only the first 70 columns and the last column (73rd) which contains the behavior label
                df = df.iloc[:, list(range(66)) + [70] + [72]]
                # Split the data based on the protective behavior label
                # Assuming the last column in df is the protective behavior label
                protective_behavior = df.iloc[:, -1]
                protective_df = df[protective_behavior == 1]
                non_protective_df = df[protective_behavior == 0]
                # Append the resulting dataframes to their respective lists
                protective_dfs.append(protective_df)
                non_protective_dfs.append(non_protective_df)

        # Concatenate all protective and non-protective dataframes
        all_protective_data = pd.concat(protective_dfs, axis=0, ignore_index=True)
        all_non_protective_data = pd.concat(non_protective_dfs, axis=0, ignore_index=True)

        # Now `all_protective_data` and `all_non_protective_data` hold the protective and non-protective data respectively
        # You can process these dataframes as needed for your analysis or save them to new .mat files
        # 保护和非保护数据个数 (77298, 68) (437247, 68)
        # 数据合并
        all_data = np.concatenate([all_protective_data,all_non_protective_data],axis=0)
        # 提取运动类型和保护性行为列
        ys = all_data[:,-2:]
        # 将特征列构建为N*V*C的形式，其中N=图个数（一条数据就可以看作一个图），V代表节点个数， C代表每个节点的特征个数
        x = all_data[:,np.newaxis,[0,22,44]]
        for node in range(2, 23):
            x_index = node - 1
            y_index = x_index + 22
            z_index = x_index + 44
            temp = all_data[:,np.newaxis,[x_index,y_index,z_index]]
            x = np.concatenate([x,temp], axis=1)
        # 将边整理为GNN要求的格式，边索引必须从0开始
       
        # 构建数据列表
        data_list = []
        # 一条数据构建为一个图
        for i in range(x.shape[0]):
            y = torch.tensor([ys[i,-1]], dtype=torch.float)   
            # 根据数据所属运动类型获得其拓扑结构
            edge_index, edge_attr = get_edge_index_and_edge_attr(ys[i,-2])
            if ys[i,-2] == 3:
                data = Data(x=torch.tensor(x[i], dtype=torch.float), edge_index=edge_index, edge_attr=edge_attr, y=y)
                data_list.append(data)
        # 将处理后的数据存储至指定根目录下
        data, slices = self.collate(data_list)
        torch.save((data, slices), self.processed_paths[0])
            
        
# 加载并分割数据集      
def load_data_and_split(dataset_root):
    # 数据打乱
    dataset = BinaryDataset(root=dataset_root)
    # 随机打乱
    dataset = dataset.shuffle()
    # 数据划分逻辑
    num_graphs = len(dataset)
    train_size = int(num_graphs * 0.8)
    # train_size = int(num_graphs * 0.01)
    
    # 创建训练集和测试集的子集索引
    train_indices = list(range(train_size))
    test_indices = list(range(train_size, num_graphs))
    
    # 使用torch_geometric.data.Subset来划分数据
    train_dataset = dataset.index_select(train_indices)
    test_dataset = dataset.index_select(test_indices)
    print("isCause:",isCause)
    print("len of train_dataset:",len(train_dataset))
    print("len of test_dataset :",len(test_dataset))
    # 创建DataLoaders
    train_loader = DataLoader(train_dataset, batch_size=bs, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=bs, shuffle=False)
    
    return train_loader, test_loader

train_loader, test_loader =  load_data_and_split(dataset_root)

Processing...


isCause: True
len of train_dataset: 10969
len of test_dataset : 2743


Done!


## 模型定义
模块参见文档 https://pytorch-geometric.readthedocs.io/en/latest/

In [5]:
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        
        self.conv1 = GCNConv(3,64)
        self.pool1 = TopKPooling(64, ratio=0.8)
        self.conv2 = GCNConv(64,64)
        self.pool2 = TopKPooling(64, ratio=0.8)
        self.conv3 = GCNConv(64,64)
        self.pool3 = TopKPooling(64, ratio=0.8)
        self.lin1 = Linear(64,64)
        self.lin2 = Linear(64,32)
        self.lin3 = Linear(32,1)
        # self.bn1 = BatchNorm1d(64)
        # self.bn2 = BatchNorm1d(32)
        self.act1 = ReLU()
        self.act2 = ReLU()

    
    def forward(self, data):
        x, edge_index, batch, edge_attr = data.x, data.edge_index, data.batch, data.edge_attr

        x = x.clone().detach().requires_grad_(True)
        x = F.relu(self.conv1(x, edge_index, edge_attr))
        # x, edge_index, _, batch, _, _ = self.pool1(x, edge_index, edge_attr, batch)
        x1 = gap(x, batch)
        
        x = F.relu(self.conv2(x, edge_index, edge_attr))
        x, edge_index, _, batch, _, _ = self.pool2(x, edge_index, edge_attr, batch)
        x2 = gap(x, batch)
        
        # x = F.relu(self.conv3(x, edge_index))
        # x, edge_index, _, batch, _, _ = self.pool3(x, edge_index, None, batch)
        # x3 = gap(x, batch)  
        
        x = x1 + x2
        x = self.lin1(x)
        x = self.act1(x)
        x = self.lin2(x)
        x = self.act2(x)
        x = F.dropout(x, p=0.5, training=self.training)
        
        # 输出值为0-1之间 （0为非保护 1为保护）
        x = torch.sigmoid(self.lin3(x)).squeeze(1)
        return x     

## 模型训练和测试

In [6]:
def train(train_loader):
    model.train()
    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        output = model(data)
        label = data.y.to(device)
        loss = crit(output, label)
        loss.backward()
        optimizer.step()

def test(loader):
    model.eval()
    y_true = []
    y_pred = []

    with torch.no_grad():  # 在评估模式下不计算梯度
        for data in loader:
            data = data.to(device)
            out = model(data)  # 前向传播
            pred = (out >= 0.5).int()

            y_true.extend(data.y.clone().detach().tolist())  # 收集真实标签
            y_pred.extend(pred.clone().detach().tolist())  # 收集预测标签
    
    accuracy = accuracy_score(y_true, y_pred)
    macro_precision = precision_score(y_true, y_pred,  pos_label=1, average='macro') 
    
    f1 = f1_score(y_true, y_pred, pos_label=1)  # Only report results for the class specified by pos_label. 考虑了非平衡
    recall = recall_score(y_true, y_pred, pos_label=1)
    return accuracy, macro_precision, f1, recall

model = Net().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=lr)
crit = torch.nn.BCELoss()
for epoch in range(1, epoch_num):
    train(train_loader)
    train_accuracy, train_macro_precision, train_f1, train_recall = test(train_loader)
    test_accuracy,  test_macro_precision,  test_f1 , test_recall  = test(test_loader)
    
    print(f'Epoch: {epoch:03d}  -------------------------- '
          f'\nTrain: accuracy: {train_accuracy:.4f}, Macro Precision: {train_macro_precision:.4f}, F1: {train_f1:.4f}, recall: {train_recall:.4f}  '
          f'\nTest : accuracy: {test_accuracy:.4f}, Macro Precision: {test_macro_precision:.4f}, F1: {test_f1:.4f}, recall: {test_recall:.4f}')

Epoch: 001  -------------------------- 
Train: accuracy: 0.8001, Macro Precision: 0.8066, F1: 0.7545, recall: 0.6905  
Test : accuracy: 0.7969, Macro Precision: 0.8012, F1: 0.7571, recall: 0.7034
Epoch: 002  -------------------------- 
Train: accuracy: 0.7693, Macro Precision: 0.7666, F1: 0.7454, recall: 0.7596  
Test : accuracy: 0.7751, Macro Precision: 0.7731, F1: 0.7556, recall: 0.7731
Epoch: 003  -------------------------- 
Train: accuracy: 0.8096, Macro Precision: 0.8090, F1: 0.7791, recall: 0.7549  
Test : accuracy: 0.8086, Macro Precision: 0.8082, F1: 0.7806, recall: 0.7569
Epoch: 004  -------------------------- 
Train: accuracy: 0.8300, Macro Precision: 0.8293, F1: 0.8040, recall: 0.7838  
Test : accuracy: 0.8349, Macro Precision: 0.8351, F1: 0.8109, recall: 0.7869
Epoch: 005  -------------------------- 
Train: accuracy: 0.8454, Macro Precision: 0.8433, F1: 0.8309, recall: 0.8543  
Test : accuracy: 0.8447, Macro Precision: 0.8432, F1: 0.8328, recall: 0.8598
Epoch: 006  --------

Epoch: 001  -------------------------- 
Train: accuracy: 0.8001, Macro Precision: 0.8066, F1: 0.7545, recall: 0.6905  
Test : accuracy: 0.7969, Macro Precision: 0.8012, F1: 0.7571, recall: 0.7034
Epoch: 002  -------------------------- 
Train: accuracy: 0.7693, Macro Precision: 0.7666, F1: 0.7454, recall: 0.7596  
Test : accuracy: 0.7751, Macro Precision: 0.7731, F1: 0.7556, recall: 0.7731
Epoch: 003  -------------------------- 
Train: accuracy: 0.8096, Macro Precision: 0.8090, F1: 0.7791, recall: 0.7549  
Test : accuracy: 0.8086, Macro Precision: 0.8082, F1: 0.7806, recall: 0.7569
Epoch: 004  -------------------------- 
Train: accuracy: 0.8300, Macro Precision: 0.8293, F1: 0.8040, recall: 0.7838  
Test : accuracy: 0.8349, Macro Precision: 0.8351, F1: 0.8109, recall: 0.7869
Epoch: 005  -------------------------- 
Train: accuracy: 0.8454, Macro Precision: 0.8433, F1: 0.8309, recall: 0.8543  
Test : accuracy: 0.8447, Macro Precision: 0.8432, F1: 0.8328, recall: 0.8598
Epoch: 006  -------------------------- 
Train: accuracy: 0.8492, Macro Precision: 0.8502, F1: 0.8243, recall: 0.7950  
Test : accuracy: 0.8502, Macro Precision: 0.8506, F1: 0.8287, recall: 0.8055
Epoch: 007  -------------------------- 
Train: accuracy: 0.8884, Macro Precision: 0.8868, F1: 0.8753, recall: 0.8805  
Test : accuracy: 0.8888, Macro Precision: 0.8875, F1: 0.8769, recall: 0.8801
Epoch: 008  -------------------------- 
Train: accuracy: 0.8496, Macro Precision: 0.8483, F1: 0.8383, recall: 0.8764  
Test : accuracy: 0.8462, Macro Precision: 0.8455, F1: 0.8368, recall: 0.8768
Epoch: 009  -------------------------- 
Train: accuracy: 0.8677, Macro Precision: 0.8657, F1: 0.8528, recall: 0.8614  
Test : accuracy: 0.8735, Macro Precision: 0.8720, F1: 0.8600, recall: 0.8639
Epoch: 010  -------------------------- 
Train: accuracy: 0.8737, Macro Precision: 0.8740, F1: 0.8546, recall: 0.8342  
Test : accuracy: 0.8753, Macro Precision: 0.8764, F1: 0.8573, recall: 0.8323
Epoch: 011  -------------------------- 
Train: accuracy: 0.8802, Macro Precision: 0.8782, F1: 0.8690, recall: 0.8930  
Test : accuracy: 0.8728, Macro Precision: 0.8714, F1: 0.8634, recall: 0.8938
Epoch: 012  -------------------------- 
Train: accuracy: 0.8933, Macro Precision: 0.8926, F1: 0.8791, recall: 0.8721  
Test : accuracy: 0.8917, Macro Precision: 0.8916, F1: 0.8779, recall: 0.8655
Epoch: 013  -------------------------- 
Train: accuracy: 0.8887, Macro Precision: 0.8868, F1: 0.8785, recall: 0.9045  
Test : accuracy: 0.8928, Macro Precision: 0.8913, F1: 0.8843, recall: 0.9100
Epoch: 014  -------------------------- 
Train: accuracy: 0.8725, Macro Precision: 0.8705, F1: 0.8589, recall: 0.8723  
Test : accuracy: 0.8735, Macro Precision: 0.8719, F1: 0.8607, recall: 0.8687
Epoch: 015  -------------------------- 
Train: accuracy: 0.8978, Macro Precision: 0.8968, F1: 0.8906, recall: 0.9356  
Test : accuracy: 0.8987, Macro Precision: 0.8985, F1: 0.8932, recall: 0.9417
Epoch: 016  -------------------------- 
Train: accuracy: 0.8976, Macro Precision: 0.8960, F1: 0.8858, recall: 0.8924  
Test : accuracy: 0.9001, Macro Precision: 0.8989, F1: 0.8893, recall: 0.8922
Epoch: 017  -------------------------- 
Train: accuracy: 0.9044, Macro Precision: 0.9029, F1: 0.8968, recall: 0.9346  
Test : accuracy: 0.9059, Macro Precision: 0.9048, F1: 0.8994, recall: 0.9344
Epoch: 018  -------------------------- 
Train: accuracy: 0.9062, Macro Precision: 0.9054, F1: 0.8940, recall: 0.8889  
Test : accuracy: 0.9074, Macro Precision: 0.9073, F1: 0.8959, recall: 0.8857
Epoch: 019  -------------------------- 
Train: accuracy: 0.9054, Macro Precision: 0.9039, F1: 0.8980, recall: 0.9361  
Test : accuracy: 0.9052, Macro Precision: 0.9043, F1: 0.8990, recall: 0.9376
Epoch: 020  -------------------------- 
Train: accuracy: 0.9055, Macro Precision: 0.9037, F1: 0.8972, recall: 0.9276  
Test : accuracy: 0.9067, Macro Precision: 0.9051, F1: 0.8987, recall: 0.9206
Epoch: 021  -------------------------- 
Train: accuracy: 0.9107, Macro Precision: 0.9088, F1: 0.9025, recall: 0.9297  
Test : accuracy: 0.9063, Macro Precision: 0.9048, F1: 0.8986, recall: 0.9230
Epoch: 022  -------------------------- 
Train: accuracy: 0.9046, Macro Precision: 0.9030, F1: 0.8938, recall: 0.9022  
Test : accuracy: 0.9059, Macro Precision: 0.9046, F1: 0.8962, recall: 0.9028
Epoch: 023  -------------------------- 
Train: accuracy: 0.9148, Macro Precision: 0.9130, F1: 0.9058, recall: 0.9211  
Test : accuracy: 0.9202, Macro Precision: 0.9189, F1: 0.9122, recall: 0.9214
Epoch: 024  -------------------------- 
Train: accuracy: 0.9216, Macro Precision: 0.9205, F1: 0.9120, recall: 0.9131  
Test : accuracy: 0.9220, Macro Precision: 0.9214, F1: 0.9130, recall: 0.9100
Epoch: 025  -------------------------- 
Train: accuracy: 0.8900, Macro Precision: 0.8880, F1: 0.8795, recall: 0.9028  
Test : accuracy: 0.8844, Macro Precision: 0.8828, F1: 0.8745, recall: 0.8947
Epoch: 026  -------------------------- 
Train: accuracy: 0.9165, Macro Precision: 0.9148, F1: 0.9092, recall: 0.9402  
Test : accuracy: 0.9223, Macro Precision: 0.9209, F1: 0.9160, recall: 0.9408
Epoch: 027  -------------------------- 
Train: accuracy: 0.9123, Macro Precision: 0.9109, F1: 0.9054, recall: 0.9434  
Test : accuracy: 0.9180, Macro Precision: 0.9167, F1: 0.9119, recall: 0.9433
Epoch: 028  -------------------------- 
Train: accuracy: 0.9101, Macro Precision: 0.9089, F1: 0.9034, recall: 0.9451  
Test : accuracy: 0.9162, Macro Precision: 0.9153, F1: 0.9107, recall: 0.9506
Epoch: 029  -------------------------- 
Train: accuracy: 0.9158, Macro Precision: 0.9139, F1: 0.9078, recall: 0.9328  
Test : accuracy: 0.9162, Macro Precision: 0.9147, F1: 0.9080, recall: 0.9198